conmon-rs/server/src/child_reaper.rs

```diff
@@ -1,169 +1,148 @@
+        let mut map = locked_grandchildren
             .lock()
             .map_err(|e| format_err!("lock grandchildren: {}", e))?;
-        let grandchild = match map.remove(&(i32::from(grandchild_pid))) {
-            Some(c) => c,
-            None => {
-                // If we have an unregistered PID, then there's nothing to do.
-                return Ok(());
-            }
-        };
-
-        debug!(
-            "PID {} associated with container {} exited with {}",
-            grandchild_pid, grandchild.id, exit_status
-        );
-        if let Err(e) = write_to_exit_paths(exit_status, grandchild.exit_paths) {
-            error!(
-                "failed to write to exit paths process for id {} :{}",
-                grandchild.id, e
-            );
-        }
+        map.retain(|_, v| v.pid == grandchild_pid);
         Ok(())
     }
 
     pub fn kill_grandchildren(&self, s: Signal) -> Result<()> {
-        for (pid, kc) in Arc::clone(&self.grandchildren)
+        for (_, grandchild) in Arc::clone(&self.grandchildren)
             .lock()
             .map_err(|e| format_err!("lock grandchildren: {}", e))?
             .iter()
         {
-            debug!("killing pid {} for container {}", pid, kc.id);
-            kill(Pid::from_raw(*pid), s)?;
+            debug!("killing pid {}", grandchild.pid);
+            kill(Pid::from_raw(grandchild.pid), s)?;
         }
         Ok(())
     }
 }
 
-#[derive(Debug, Default)]
-pub struct ChildReaper {
-    grandchildren: Arc<Mutex<HashMap<i32, ReapableChild>>>,
-    wait_lock: Arc<Mutex<bool>>,
-}
-
-#[derive(Debug, Getters)]
+#[derive(Default, Debug, Clone)]
 pub struct ReapableChild {
-    #[getset(get)]
-    id: String,
-    #[getset(get)]
-    exit_paths: Vec<PathBuf>,
+    pub exit_paths: Vec<PathBuf>,
+    pub pid: i32,
 }
 
 impl ReapableChild {
     pub fn from_child(child: &Child) -> Self {
         Self {
-            id: child.id.clone(),
+            pid: child.pid,
             exit_paths: child.exit_paths.clone(),
         }
     }
+
+    fn watch(&self) -> tokio::sync::oneshot::Receiver<()> {
+        let exit_paths = self.exit_paths.clone();
+        let pid = self.pid;
+        let (tx, rx) = tokio::sync::oneshot::channel();
+
+        tokio::task::spawn_blocking(move || {
+            let wait_status = waitpid(Pid::from_raw(pid), None);
+            match wait_status {
+                Ok(status) => {
+                    if let WaitStatus::Exited(_, exit_status) = status {
+                        let _ = write_to_exit_paths(exit_status, &exit_paths);
+                    }
+                }
+                Err(err) => {
+                    if err != nix::errno::Errno::ECHILD {
+                        error!("caught error in reading for sigchld {}", err);
+                    }
+                }
+            };
+            if tx.send(()).is_err() {
+                error!("the receiver dropped the watch")
+            }
+        });
+        rx
```

Comment:
whereas the refactoring of child reaping and pid watching should be in a separate commit describing the reasoning